File: linker/lib/memory_parser.rb
require File.dirname(__FILE__) + '/parser'

class MemoryParser < Parser

  def parse
    @char = reader.next
    MemoryMap.instance.modules.each do |program_module|
      detect_uses(program_module)
      detect_instructions(program_module)
    end
    reader.file.rewind
    
    MemoryMap.validate!
  end
  
private
  def detect_uses(program_module)
    skip_symbols
    parse_number.times do |i|             # Start of the use list, parse_number is NU
      program_module.uses << parse_word   # External symbols, parse_number is X21, X31, etc.
    end
  end
  
  def detect_instructions(program_module)
    parse_number.times do |i|             # Start of instr set, parse_number is NT
      type = parse_word                   # Instruction type, parse_word is I, R, E, or A
      word = parse_number                 # Instr, parse_number is Opcode + Address
      
      program_module.create_instruction(type, word)
    end
    program_module.map_instructions
  end
  
  def skip_symbols
    parse_number.times do |i| 
      parse_word
      parse_number
    end
  end
  
end

File: java/linker/src/lib/MemoryParser.java
package lib;

import java.io.IOException;

/**
 * User: Ross Kaffenberger
 * Date: Sep 22, 2009
 * Time: 11:36:22 PM
 */
public class MemoryParser extends Parser {
    public MemoryParser(Linker linker) {
        super(linker);
    }

    public void parse() throws IOException {
        readNextCharString();
        for (ProgramModule module : MemoryMap.getInstance().getModules()) {
            detectUses(module);
            detectInstructions(module);
        }
        getReader().rewind();

        MemoryMap.validate();
    }

    private void detectInstructions(ProgramModule module) throws IOException {
        int nt = parseNumber();
        for (int i = 1; i <= nt; i++) {
            String type = parseWord();
            int word = parseNumber();

            module.createInstruction(type, word);
        }
        module.mapInstructions();
    }

    private void skipSymbols() throws IOException {
        int times = parseNumber();
        for (int i = 1; i <= times; i++) {
            parseWord();
            parseNumber();
        }

    }

    private void detectUses(ProgramModule module) throws IOException {
        skipSymbols();
        int nu = parseNumber();
        for (int i = 1; i <= nu; i++) {
            module.addUse(parseWord());
        }
    }
}
